sonic_y_cable/y_cable.py

```diff
@@ -1097,11 +1129,104 @@ def check_if_nic_lanes_active(physical_port):
 
     if platform_chassis is not None:
         res = platform_chassis.get_sfp(physical_port).read_eeprom(curr_offset, 1)
-        y_cable_validate_read_data(res, 1, physical_port, "nic lanes active")
+        if y_cable_validate_read_data(res, 1, physical_port, "nic lanes active") == EEPROM_READ_DATA_INVALID:
+            return EEPROM_ERROR
         result = res[0]
 
     else:
         helper_logger.log_error("platform_chassis is not loaded, failed to get NIC lanes active")
         return -1
 
     return result
+
+
+def get_firmware_version(physical_port, target):
+
+    data = bytearray(FIRMWARE_INFO_PAYLOAD_SIZE)
+
+    if platform_chassis is not None:
+        for byteIdx in range(0, FIRMWARE_INFO_PAYLOAD_SIZE):
+            curr_offset = 0xfc * 128 + 128 + byteIdx
+            readOut = platform_chassis.get_sfp(physical_port).read_eeprom(curr_offset, 1)
+            if y_cable_validate_read_data(readOut, 1, physical_port, "firmware info") == EEPROM_READ_DATA_INVALID:
+                return EEPROM_ERROR
+            data[byteIdx] = readOut[0]
+    else:
+        helper_logger.log_error("platform_chassis is not loaded, failed to get NIC lanes active")
+        return -1
+
+    result = []
+    NUM_MCU_SIDE = 3
+    Addr = target * (FIRMWARE_INFO_PAYLOAD_SIZE / NUM_MCU_SIDE)
+    baseAddr = int(Addr)
+    rev_major_slot1 = struct.unpack_from('<B', data[0 + baseAddr:  1 + baseAddr])[0]
+    rev_minor_slot1 = struct.unpack_from('<B', data[2 + baseAddr:  3 + baseAddr])[0]
+    rev_build_lsb_slot1 = struct.unpack_from('<B', data[4 + baseAddr:  5 + baseAddr])[0]
+    rev_build_msb_slot1 = struct.unpack_from('<B', data[5 + baseAddr:  6 + baseAddr])[0]
+    rev_major_slot2 = struct.unpack_from('<B', data[1 + baseAddr:  2 + baseAddr])[0]
+    rev_minor_slot2 = struct.unpack_from('<B', data[3 + baseAddr:  4 + baseAddr])[0]
+    rev_build_lsb_slot2 = struct.unpack_from('<B', data[6 + baseAddr:  7 + baseAddr])[0]
+    rev_build_msb_slot2 = struct.unpack_from('<B', data[7 + baseAddr:  8 + baseAddr])[0]
+    slotStatus = struct.unpack_from('<B', data[8 + baseAddr:  9 + baseAddr])[0]
+
+    if (rev_major_slot1 == 0 and rev_minor_slot1 == 0 and rev_build_lsb_slot1 == 0 and rev_build_msb_slot1 == 0 and rev_major_slot2 == 0 and rev_minor_slot2 == 0 and rev_build_lsb_slot2 == 0 and rev_build_msb_slot2 == 0):
+        return -1
```

Comment:
Prefer not to have different return value types (here list/int). Suggest either returning `None` on failure.

Reply:
fixed